File: Rotting Oranges.py
#https://leetcode.com/problems/rotting-oranges/?envType=study-plan-v2&envId=leetcode-75

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        positionI=-1
        positionJ=-1
        maxii=-1
        queue=[] #TO STORE THE POSITION
        #FIND ALL 2 IN THE GRID
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j]==2:
                    positionI=i
                    positionJ=j
                    queue.append([positionI, positionJ])
                    grid[i][j]=5 #let level 0 be 5
        ########################################
        #no 2 found
        if positionI==-1 and positionJ==-1:
            for i in grid:
                if 1 in i:
                    return -1
            return 0
        ####################################
        else:
            #BFS
            #let level 0 be 5
            maxii=5
            while queue:
                val=queue.pop(0)
                i,j=val[0],val[1]

                if i-1>-1 and grid[i-1][j]==1:
                    grid[i-1][j]=grid[i][j]+1 #MARKING THE LEVEL +1
                    queue.append([i-1,j]) #inserting in queue
                    maxii=max(maxii,grid[i-1][j])
                if i+1<len(grid) and grid[i+1][j]==1:
                    grid[i+1][j]=grid[i][j]+1
                    queue.append([i+1,j])
                    maxii=max(maxii,grid[i+1][j])
                if j-1>=0 and grid[i][j-1]==1:
                    grid[i][j-1]=grid[i][j]+1
                    queue.append([i,j-1])
                    maxii=max(maxii,grid[i][j-1])
                if j+1<len(grid[0]) and grid[i][j+1]==1:
                    grid[i][j+1]=grid[i][j]+1
                    queue.append([i,j+1])
                    maxii=max(maxii,grid[i][j+1])
                
            #IF STILL 1 present in grid
            for i in grid:
                if 1 in i:
                    return -1
            return maxii-5
        
    

File: Rotting Oranges.cpp
//https://leetcode.com/problems/rotting-oranges/?envType=problem-list-v2&envId=interview-instance-ii

class Solution {
public:
    
    int orangesRotting(vector<vector<int>>& grid) {
        vector<vector<int>> time(grid.size(),vector<int> (grid[0].size(),1e9));
        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[0].size();j++){
                if(grid[i][j]==2){
                    time[i][j]=0;
                    queue<pair<int,int>> q;
                    q.push({i,j});
                    while(!q.empty()){
                        auto node=q.front();
                        q.pop();
                        int x=node.first, y=node.second;
                        if(x-1>-1 && grid[x-1][y]==1){
                            if(time[x-1][y]>time[x][y]+1){
                                q.push({x-1,y});
                                time[x-1][y]=time[x][y]+1;
                            }
                        }
                        if(y-1>-1 && grid[x][y-1]==1){
                            if(time[x][y-1]>time[x][y]+1){
                                q.push({x,y-1});
                                time[x][y-1]=time[x][y]+1;
                            }
                        }
                        if(x+1<grid.size() && grid[x+1][y]==1 && time[x+1][y]>time[x][y]+1){
                            q.push({x+1,y});
                            time[x+1][y]=time[x][y]+1;
                        }
                        if(y+1<grid[0].size() && grid[x][y+1]==1 && time[x][y+1]>time[x][y]+1){
                            q.push({x,y+1});
                            time[x][y+1]=time[x][y]+1;
                        }
                    }
                }
            }
        }
        int result=0;
        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[0].size();j++){
                if(grid[i][j]==1 && time[i][j]==1e9) return -1;
                if(grid[i][j]==1 || grid[i][j]==2) result=max(result,time[i][j]);
            }
        }
        return result;
    }
};
